File: 01-Arrays/implementations/python/DynamicArray.py
class DynamicArray:

    def __init__(self):
        self.capacity = 2
        self.size = 0
        self.array = [None] * self.capacity

    def add(self, value):

        if self.size == self.capacity:
            self._grow()

        self.array[self.size] = value
        self.size += 1

    def insert(self, index, value):

        if self.size == self.capacity:
            self._grow()

        for i in range(self.size, index, -1):
            self.array[i] = self.array[i - 1]

        self.array[index] = value
        self.size += 1

    def remove(self, index):

        for i in range(index, self.size - 1):
            self.array[i] = self.array[i + 1]

        self.array[self.size - 1] = None
        self.size -= 1

        if self.size <= self.capacity // 4 and self.capacity > 2:
            self._shrink()

    def get(self, index):
        return self.array[index]

    def set(self, index, value):
        self.array[index] = value

    def _grow(self):

        self.capacity *= 2

        new_array = [None] * self.capacity

        for i in range(self.size):
            new_array[i] = self.array[i]

        self.array = new_array

    def _shrink(self):

        self.capacity //= 2

        new_array = [None] * self.capacity

        for i in range(self.size):
            new_array[i] = self.array[i]

        self.array = new_array

    def display(self):

        print(self.array)
        print(f"Size = {self.size}")
        print(f"Capacity = {self.capacity}")

File: 01-Arrays/implementations/java/DynamicArray.java
public class DynamicArray {

    private int size;
    private int capacity;
    private int[] array;

    public DynamicArray() {
        capacity = 2;
        array = new int[capacity];
    }

    public void add(int value) {

        if (size == capacity) {
            grow();
        }

        array[size++] = value;
    }

    public int get(int index) {

        if (index >= size)
            throw new IndexOutOfBoundsException();

        return array[index];
    }

    public void set(int index, int value) {

        if (index >= size)
            throw new IndexOutOfBoundsException();

        array[index] = value;
    }

    public void insert(int index, int value) {

        if (size == capacity)
            grow();

        for (int i = size; i > index; i--) {
            array[i] = array[i - 1];
        }

        array[index] = value;
        size++;
    }

    public void remove(int index) {

        if (index >= size)
            throw new IndexOutOfBoundsException();

        for (int i = index; i < size - 1; i++) {
            array[i] = array[i + 1];
        }

        size--;

        if (size <= capacity / 4 && capacity > 2) {
            shrink();
        }
    }

    private void grow() {

        capacity *= 2;

        int[] newArray = new int[capacity];

        for (int i = 0; i < size; i++) {
            newArray[i] = array[i];
        }

        array = newArray;
    }

    private void shrink() {

        capacity /= 2;

        int[] newArray = new int[capacity];

        for (int i = 0; i < size; i++) {
            newArray[i] = array[i];
        }

        array = newArray;
    }

    public int size() {
        return size;
    }

    public int capacity() {
        return capacity;
    }

    public void print() {

        System.out.print("[ ");

        for (int i = 0; i < capacity; i++) {

            if (i < size)
                System.out.print(array[i] + " ");
            else
                System.out.print("_ ");
        }

        System.out.println("]");
    }
}
